File: src/momentarily/arrivals.py
"""Per-stop next-N arrivals — the Python mirror of worker/src/arrivals.ts.

The Worker derives the published `arrivals` surface; this reproduces the same
derivation so the two stay in lockstep (pinned by tests/fixtures/parity_arrivals
.json). Same rules, not a second set: fold the express variant to the base
route, key by the stop id incl. its direction suffix (so N/S key separately),
drop SKIPPED / timeless / past / beyond-horizon rows, one entry per trip per
stop (first occurrence wins), soonest first, capped at N.

Times are the feed's own absolute POSIX seconds; seconds_away is time - now.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TypedDict

# Don't publish a countdown further out than this — past an hour the times are
# speculative and just pad the list. Matches worker/src/arrivals.ts.
HORIZON_SECONDS = 60 * 60
# The next few trains is all a countdown board shows.
MAX_ARRIVALS = 6

# schedule_relationship enum: 0=SCHEDULED, 1=SKIPPED, 2=NO_DATA, 3=UNSCHEDULED.
SKIPPED = 1


@dataclass(frozen=True)
class StopTimeLite:
    """One decoded StopTimeUpdate row — mirror of the TS StopTimeLite."""

    stop_id: str  # GTFS stop id incl. direction suffix; '' when absent
    arrival: int | None
    departure: int | None
    schedule_relationship: int = 0


@dataclass(frozen=True)
class TripLite:
    """A decoded trip carrying its stop times — the arrivals-relevant slice of
    the TS TripLite. `direction` is the NYCT extension enum (1=N, 3=S) or None."""

    route_id: str
    trip_id: str
    direction: int | None = None
    stop_times: list[StopTimeLite] = field(default_factory=list[StopTimeLite])


class ArrivalDict(TypedDict):
    """One entry in the per-stop arrivals list — the JSON shape schema.Arrival
    validates and worker/src/arrivals.ts Arrival mirrors."""

    route: str
    eta_epoch: int
    seconds_away: int
    trip_id: str | None


def base_route(route_id: str) -> str:
    """Fold the express variant to its base route (6X -> 6). Mirrors
    worker/src/trip_updates.ts baseRoute."""
    return route_id[:-1] if route_id.endswith("X") else route_id


def derive_arrivals(trips: list[TripLite], now: int) -> dict[str, list[ArrivalDict]]:
    """Fold decoded trips into per-stop arrivals, soonest first. One pass over
    every row, then a sort + truncate per stop. See the module docstring for the
    drop rules; each returned entry is {route, eta_epoch, seconds_away, trip_id}.
    """
    out: dict[str, list[ArrivalDict]] = {}
    seen: dict[str, set[str]] = {}
    horizon = now + HORIZON_SECONDS

    for trip in trips:
        route = base_route(trip.route_id)
        for st in trip.stop_times:
            if st.schedule_relationship == SKIPPED:
                continue
            time = st.arrival if st.arrival is not None else st.departure
            if time is None:
                continue
            stop_id = st.stop_id
            if stop_id == "":
                continue
            if time < now or time > horizon:
                continue

            stop_seen = seen.get(stop_id)
            if stop_seen is None:
                stop_seen = set[str]()
                seen[stop_id] = stop_seen
                out[stop_id] = []
            # An anonymous trip cannot be told from another anonymous trip, so
            # it is never deduped: two trains the feed left unnamed are two
            # arrivals, not one.
            if trip.trip_id != "":
                if trip.trip_id in stop_seen:
                    continue
                stop_seen.add(trip.trip_id)

            out[stop_id].append(
                ArrivalDict(
                    route=route,
                    eta_epoch=time,
                    seconds_away=time - now,
                    trip_id=trip.trip_id if trip.trip_id != "" else None,
                )
            )

    for entries in out.values():
        entries.sort(key=lambda a: a["eta_epoch"])
        del entries[MAX_ARRIVALS:]
    return out

File: worker/src/arrivals.ts
/**
 * Per-stop next-N arrivals, derived from the decoded GTFS-RT trip-updates.
 *
 * gtfsrt.ts materializes each StopTimeUpdate (stop_id incl. direction suffix,
 * absolute POSIX arrival/departure times, schedule_relationship). This folds
 * those rows into a ranked list per stop id: the next few trains due, soonest
 * first, one entry per physical trip. Route is folded to its base (6X -> 6)
 * exactly as the service metric does; direction is carried by the stop id's own
 * suffix (Q05S / 626N), so northbound and southbound key separately.
 *
 * No schedule join and no delay math: eta_epoch is the feed's own time,
 * seconds_away is just time - now. Absolute times, carried through untouched.
 */

import type { TripLite } from './gtfsrt';
import { baseRoute } from './trip_updates';

/** One upcoming train at a stop. `trip_id` is the NYCT run id, null when the
 * feed omitted it. Times are absolute POSIX seconds; seconds_away = time - now. */
export interface Arrival {
  route: string;
  eta_epoch: number;
  seconds_away: number;
  trip_id: string | null;
}

/** Keyed by GTFS stop id incl. direction suffix (e.g. 'Q05S'), soonest first. */
export type Arrivals = Record<string, Arrival[]>;

// Don't publish a countdown further out than this — past an hour the trip-
// update times are speculative and just pad the list.
const HORIZON_SECONDS = 60 * 60;
// Most a rider reads off a countdown board is the next few trains.
const MAX_ARRIVALS = 6;

/**
 * Fold decoded trips (across all fetched feeds) into per-stop arrivals.
 *
 * One pass over every stop-time row, pushing into a per-stop array, then a
 * sort + truncate per stop at the end. A row is dropped when it is SKIPPED,
 * carries neither an arrival nor a departure time, is keyed to no stop, is
 * already in the past, or sits beyond the horizon. One entry per trip per stop
 * — the first occurrence wins, so a trip listed twice for a stop counts once.
 */
export function deriveArrivals(trips: TripLite[], now: number): Arrivals {
  const out: Arrivals = {};
  // Per-stop set of trip ids already counted, for the one-entry-per-trip rule.
  const seen = new Map<string, Set<string>>();
  const horizon = now + HORIZON_SECONDS;

  for (const t of trips) {
    const route = baseRoute(t.routeId);
    for (const st of t.stopTimes) {
      if (st.scheduleRelationship === 1) continue; // SKIPPED
      const time = st.arrival ?? st.departure;
      if (time === null) continue;
      const stopId = st.stopId;
      if (stopId === '') continue;
      if (time < now || time > horizon) continue;

      let stopSeen = seen.get(stopId);
      if (stopSeen === undefined) {
        stopSeen = new Set<string>();
        seen.set(stopId, stopSeen);
        out[stopId] = [];
      }
      // An anonymous trip cannot be told from another anonymous trip, so it is
      // never deduped: two trains the feed left unnamed are two arrivals.
      if (t.tripId !== '') {
        if (stopSeen.has(t.tripId)) continue; // one entry per trip per stop
        stopSeen.add(t.tripId);
      }

      out[stopId]!.push({
        route,
        eta_epoch: time,
        seconds_away: time - now,
        trip_id: t.tripId === '' ? null : t.tripId,
      });
    }
  }

  for (const stopId of Object.keys(out)) {
    const list = out[stopId]!;
    list.sort((a, b) => a.eta_epoch - b.eta_epoch);
    if (list.length > MAX_ARRIVALS) list.length = MAX_ARRIVALS;
  }
  return out;
}
